docs/modules/architecture/pages/compact-binary-specification.adoc

```diff
@@ -0,0 +1,719 @@
+
+To distinguish different types supported in the wire-level format, a new enum called `FieldKind` is introduced which has different ids for different types compared to the `FieldType` enum.
+
+The `FieldType` enum has different ways to represent ids for types and their arrays and has some unfortunate public methods which would not fit the new format. (for example `FieldType#DATE` returns `getTypeSize` wrong for the new format). It is desired to have a new enum that does not have those problems and is more open to type additions in the future.
+
+Note that, the new `FieldKind` enum with those new ids will be used in the `GenericRecord` API for both `Portable` and the new format. `FieldType` and `Portable` will be deprecated in the future.
+
+=== Unsigned Integers
+
+There is support for unsigned integer types. The representation of unsigned integers is done with the smallest signed integer type that can represent it (except for `u64`, since there is no such type). So, basically,
+
+* `u8` is represented by → `i16`
+* `u16` is represented by → `i32`
+* `u32` is represented by → `i64`
+* `u64` is represented by → `BigInteger`
+
+=== Enums
+
+There is no support for enums in the wire-level format. The representation of enums is left to the user. 
+
+=== Nullable Primitives
+
+There is support for nullable primitive types as they can be useful in SQL and might play nicely with languages that have a concept to represent them. For example, a C# user might use `Nullable<PrimitiveType>`, or Java users might use the class representation of primitive types.
+
+They are implemented as variable-sized types. The `null` values of such types are represented exactly as `null` variable-sized fields, with the offset of `-1` and no data.
+
+== Header
+
+The partition hash and the type id are common for all serialization methods supported by Hazelcast. Therefore, the new format is no exception and every serialized object has a header in addition to the payload on the wire.
+
+[cols="1,1,1"]
+|===
+|Name |Type |Description
+|Partition hash |i32 |`BIG_ENDIAN` integer, used for key objects. Not applicable to value objects.
+|Type id |i32 |`BIG_ENDIAN` integer that determines the serializer to be used. -55 for compact.
+|===
+
+== Var-Size Objects
+
+In this section, how a user-defined type is represented at the wire level is described. Consists of `Header`, `Data`, and `Offsets` sections in this order.
+
+=== Header Section
+
+[cols="1,1,1"]
+|===
+|Name |Type |Description
+|Schema id |i64 |	
+Hash of the schema.
+|Data length |i32 |Length of the DATA SECTION below.
+|===
+
+=== Data Section
+
+[cols="1,1"]
+|===
+|Name |Description
+|Fixed-size Fields | Offsets of these fields will be deduced from the schema
+|Variable-size Fields | 
+|===
+
+=== Offsets Section
+
+[cols="1,1,1"]
+|===
+|Name |Type |Description
+|Variable-Size FieldOffset index 0 |u8/u16/i32 |The index of a field offset is written in the Schema. Offsets of variable length fields. -1 for null
```

Comment:
I believe we should either duplicate the description to all rows, or remove that column altogether here and explain the null case below